src/Features/CSharp/Portable/CodeRefactorings/InvertIf/InvertIfCodeRefactoringProvider.cs

```diff
@@ -1,332 +1,123 @@
-                                SyntaxKind.BarBarToken,
-                                binaryExpression.OperatorToken.TrailingTrivia),
-                            Negate(binaryExpression.Right, semanticModel, cancellationToken));
-
-                        return result
-                            .Parenthesize()
-                            .WithLeadingTrivia(binaryExpression.GetLeadingTrivia())
-                            .WithTrailingTrivia(binaryExpression.GetTrailingTrivia());
-                    }
-
-                case SyntaxKind.TrueLiteralExpression:
-                    {
-                        var literalExpression = (LiteralExpressionSyntax)expression;
-                        return SyntaxFactory.LiteralExpression(
-                            SyntaxKind.FalseLiteralExpression,
-                            SyntaxFactory.Token(
-                                literalExpression.Token.LeadingTrivia,
-                                SyntaxKind.FalseKeyword,
-                                literalExpression.Token.TrailingTrivia));
-                    }
+            return operatorToken.Kind() == SyntaxKind.AmpersandAmpersandToken;
+        }
 
-                case SyntaxKind.FalseLiteralExpression:
-                    {
-                        var literalExpression = (LiteralExpressionSyntax)expression;
-                        return SyntaxFactory.LiteralExpression(
-                            SyntaxKind.TrueLiteralExpression,
-                            SyntaxFactory.Token(
-                                literalExpression.Token.LeadingTrivia,
-                                SyntaxKind.TrueKeyword,
-                                literalExpression.Token.TrailingTrivia));
-                    }
-            }
+        internal override bool IsConditionalOr(SyntaxToken operatorToken)
+        {
+            return operatorToken.Kind() == SyntaxKind.BarBarToken;
+        }
 
-            // Anything else we can just negate by adding a ! in front of the parenthesized expression.
-            // Unnecessary parentheses will get removed by the simplification service.
-            return SyntaxFactory.PrefixUnaryExpression(
-                SyntaxKind.LogicalNotExpression,
-                expression.Parenthesize());
+        internal override bool IsFalseLiteralExpression(SyntaxNode expression)
+        {
+            return ((LiteralExpressionSyntax)expression).IsKind(SyntaxKind.FalseLiteralExpression);
         }
 
-        private class MyCodeAction : CodeAction.DocumentChangeAction
+        internal override bool IsTrueLiteralExpression(SyntaxNode expression)
         {
-            public MyCodeAction(string title, Func<CancellationToken, Task<Document>> createChangedDocument) :
-                base(title, createChangedDocument)
-            {
-            }
+            return ((LiteralExpressionSyntax)expression).IsKind(SyntaxKind.TrueLiteralExpression);
```

Comment:
1. consider following the pattern with other refactoring/fixproviders, and make the base type generic and parametrized on these types.  That way you don't need all these casts in your code.
2. you don't need the cast at all, you can just do ```expression.IsKind(...)```
3. You could just make these SyntaxFacts.